src/topology/sheaves/stalks.lean

```diff
@@ -72,69 +86,172 @@ The germ of a section of a presheaf over an open at a point of that open.
 def germ (F : X.presheaf C) {U : opens X} (x : U) : F.obj (op U) ⟶ stalk F x :=
 colimit.ι ((open_nhds.inclusion x.1).op ⋙ F) (op ⟨U, x.2⟩)
 
-/-- For a `Type` valued presheaf, every point in a stalk is a germ. -/
-lemma germ_exist (F : X.presheaf (Type v)) (x : X) (t : stalk F x) :
-  ∃ (U : opens X) (m : x ∈ U) (s : F.obj (op U)), F.germ ⟨x, m⟩ s = t :=
+@[simp, elementwise]
+lemma germ_res (F : X.presheaf C) {U V : opens X} (i : U ⟶ V) (x : U) :
+  F.map i.op ≫ germ F x = germ F (i x : V) :=
+let i' : (⟨U, x.2⟩ : open_nhds x.1) ⟶ ⟨V, (i x : V).2⟩ := i in
+colimit.w ((open_nhds.inclusion x.1).op ⋙ F) i'.op
+
+/--
+A morphism from the stalk of `F` at `x` to some object `Y` is completely determined by its
+composition with the `germ` morphisms.
+-/
+lemma stalk_hom_ext (F : X.presheaf C) {x} {Y : C} {f₁ f₂ : F.stalk x ⟶ Y}
+  (ih : ∀ (U : opens X) (hxU : x ∈ U), F.germ ⟨x, hxU⟩ ≫ f₁ = F.germ ⟨x, hxU⟩ ≫ f₂) : f₁ = f₂ :=
+colimit.hom_ext $ λ U, by { op_induction U, cases U with U hxU, exact ih U hxU }
+
+@[simp, reassoc, elementwise]
+lemma stalk_functor_map_germ {F G : X.presheaf C} (U : opens X) (x : U)
+  (f : F ⟶ G) : germ F x ≫ (stalk_functor C x.1).map f = f.app (op U) ≫ germ G x :=
+colimit.ι_map (whisker_left ((open_nhds.inclusion x.1).op) f) (op ⟨U, x.2⟩)
+
+variables (C)
+
+/--
+For a presheaf `F` on a space `X`, a continuous map `f : X ⟶ Y` induces a morphisms between the
+stalk of `f _ * F` at `f x` and the stalk of `F` at `x`.
+-/
+def stalk_pushforward (f : X ⟶ Y) (F : X.presheaf C) (x : X) : (f _* F).stalk (f x) ⟶ F.stalk x :=
 begin
-  obtain ⟨U, s, e⟩ := types.jointly_surjective _ (colimit.is_colimit _) t,
-  revert s e,
-  rw [(show U = op (unop U), from rfl)],
-  generalize : unop U = V, clear U,
-  cases V with V m,
-  intros s e,
-  exact ⟨V, m, s, e⟩,
+  -- This is a hack; Lean doesn't like to elaborate the term written directly.
+  transitivity,
+  swap,
+  exact colimit.pre _ (open_nhds.map f x).op,
+  exact colim.map (whisker_right (nat_trans.op (open_nhds.inclusion_map_iso f x).inv) F),
```

Comment:
A slighter terser hack:
```suggestion
  refine _ ≫ colimit.pre _ (open_nhds.map f x).op,
  exact colim.map (whisker_right (nat_trans.op (open_nhds.inclusion_map_iso f x).inv) F),
```
I'll think some more about what's going on here.

Reply:
Haha... I had forgotten that I wrote this, and you're just moving it around now. :-) Stet, I guess.

Reply:
Refactoring 2-year-old hacks is outside the scope of this PR ;)
I never really deciphered what's going on here. Maybe I'm going to try as well, just for fun.